lib/src/test/kotlin/graphql/nadel/engine/blueprint/NadelVirtualTypeBlueprintFactoryTest.kt

```diff
@@ -0,0 +1,62 @@
+package graphql.nadel.engine.blueprint
+
+import graphql.schema.idl.SchemaGenerator
+import org.junit.jupiter.api.Test
+import kotlin.test.assertTrue
+
+class NadelVirtualTypeBlueprintFactoryTest {
+    @Test
+    fun `creates a mapping for one scalar`() {
+        val schema = SchemaGenerator.createdMockedSchema(
+            // language=GraphQL
+            """
+                directive @hydrated(field: String!) on FIELD_DEFINITION
+
+                type Query {
+                    virtualEcho: VirtualEcho @hydrated(field: "echo")
+                    echo: BackingEcho
+                }
+                type VirtualEcho {
+                    echo: String
+                    info: VirtualEchoInfo
+                }
+                type VirtualEchoInfo {
+                    cursor: String
+                }
+                type BackingEcho {
+                    echo: String
+                    info: BackingEchoInfo
+                }
+                type BackingEchoInfo {
+                    cursor: String
+                    something: Boolean
+                }
+            """.trimIndent()
+        )
+
+        val virtualTypeContext = NadelVirtualTypeBlueprintFactory()
+            .makeVirtualTypeContext(
+                schema,
+                containerType = schema.queryType,
+                virtualFieldDef = schema.queryType.getField("virtualEcho"),
+            )
+
+        assertTrue(virtualTypeContext != null)
+
+        assertTrue(virtualTypeContext.virtualFieldContainer.name == "Query")
+        assertTrue(virtualTypeContext.virtualField.name == "virtualEcho")
+
+        assertTrue(
+            virtualTypeContext.virtualTypeToBackingType == mapOf(
+                "VirtualEcho" to "BackingEcho",
+                "VirtualEchoInfo" to "BackingEchoInfo",
+            ),
+        )
```

Comment:
Question -

So when we make a virtual type - we say that all fields in other types get made "virtual" as well - eg the shape of virtual type is actually the backing type shapes

What happens if the types reference themselves

```
 type VirtualEcho {
                    echo: String
                    info: VirtualEchoInfo
                    selfReference : VirtualEcho.  # what happens here
                }
```

Reply:
will it go into a infinite loop and if not can we have a test for that?

Reply:
There's code to protect against that, I've created a test for that though.

Branch is https://github.com/atlassian-labs/nadel/tree/remaining-arguments

Reply:
There's code to protect against that, I've created a test for that though.